### PR TITLE
Multiple Dex Selection

### PR DESCRIPTION
**Pull request by Drypto13**

### CODE REVIEW

**Review by Dpalazue — APPROVED** [review]

File: contracts/interfaces/ICurve.sol [hunk 1]
```diff
@@ -0,0 +1,32 @@
+pragma solidity 0.5.17;
```
Comment (RomanHiden) [inline contracts/interfaces/ICurve.sol:1]: make this a range up to 8.something, easier to upgrade later

File: contracts/interfaces/ICurvePoolRegistration.sol [hunk 2]
```diff
@@ -0,0 +1,15 @@
+pragma solidity 0.5.17;
```
Comment (RomanHiden) [inline contracts/interfaces/ICurvePoolRegistration.sol:1]: make a version range, exactly how in IBZRx.sol

File: contracts/interfaces/uniswapQuoter.sol [hunk 3]
```diff
@@ -0,0 +1,15 @@
+pragma solidity 0.5.17;
```
Comment (RomanHiden) [inline contracts/interfaces/uniswapQuoter.sol:1]: this file upper case UniswapQuoter

File: contracts/modules/LoanClosings/LoanClosings.sol [hunk 4]
```diff
@@ -90,7 +90,7 @@ contract LoanClosings is LoanClosingsBase {
         address receiver,
         uint256 swapAmount, // denominated in collateralToken
         bool returnTokenIsCollateral, // true: withdraws collateralToken, false: withdraws loanToken
-        bytes memory /*loanDataBytes*/) // for future use
+        bytes memory loanDataBytes) // for future use
```
Comment (RomanHiden) [inline contracts/modules/LoanClosings/LoanClosings.sol:93]: please describe structure of loanDataBytes in our IBZRx interface file in detail. current features and how to extend to more futures

File: contracts/modules/LoanOpenings/LoanOpenings.sol [hunk 5]
```diff
@@ -112,9 +112,9 @@ contract LoanOpenings is State, LoanOpeningsEvents, VaultController, InterestUse
         uint256 loanTokenSent,
         uint256 collateralTokenSent,
         uint256 interestRate,
-        uint256 newPrincipal)
+        uint256 newPrincipal,
```
Comment (RomanHiden) [inline contracts/modules/LoanOpenings/LoanOpenings.sol:115]: why we need those changes in this pull request?

File: contracts/modules/SwapsExternal/SwapsExternal.sol [hunk 6]
```diff
@@ -115,15 +115,16 @@ contract SwapsExternal is State, VaultController, SwapsUser, PausableGuardian {
...
         external
-        view
         returns (uint256)
     {
         return _swapsExpectedReturn(
             sourceToken,
             destToken,
-            sourceTokenAmount
+            sourceTokenAmount,
+			payload
```
Comment (RomanHiden) [inline contracts/modules/SwapsExternal/SwapsExternal.sol:127]: format

File: contracts/modules/LoanOpenings/LoanOpenings.sol [hunk 7]
```diff
@@ -112,9 +112,9 @@ contract LoanOpenings is State, LoanOpeningsEvents, VaultController, InterestUse
...
+		bytes calldata payload)
```
Comment (RomanHiden) [inline contracts/modules/LoanOpenings/LoanOpenings.sol:116]: describe format of payload.

File: contracts/swaps/SwapsUser.sol [hunk 8]
```diff
@@ -160,23 +162,44 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
...
+				addrs[0], // sourceToken
+				addrs[1], // destToken
+				addrs[2], // receiverAddress
+				addrs[3], // returnToSenderAddress
+				vals[0],  // minSourceTokenAmount
+				vals[1],  // maxSourceTokenAmount
+				vals[2],  // requiredDestTokenAmount
+				loanDataBytes
+			);
+		}else{
```
Comment (RomanHiden) [inline contracts/swaps/SwapsUser.sol:185]: format

File: contracts/swaps/SwapsUser.sol [hunk 9]
```diff
@@ -160,23 +162,44 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
...
-            addrs[3], // returnToSenderAddress
-            vals[0],  // minSourceTokenAmount
-            vals[1],  // maxSourceTokenAmount
-            vals[2]   // requiredDestTokenAmount
-        );
+		bytes memory data;
+		address swapImplAddress;
+		if(loanDataBytes.length==0){
+			swapImplAddress = IDexRecords(swapsImpl).retrieveDexAddress(1); //if nothing specified, default to first dex option available. ensure it does not require any input data or else this will break
+			data = abi.encodeWithSelector(
```
Comment (RomanHiden) [inline contracts/swaps/SwapsUser.sol:174]: I think encodeWithSelector can be move out of if else. but check if its worth doing it

File: contracts/swaps/SwapsUser.sol [hunk 10]
```diff
@@ -192,22 +215,32 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
...
     {
         uint256 tradingFee = _getTradingFee(sourceTokenAmount);
         if (tradingFee != 0) {
             sourceTokenAmount = sourceTokenAmount
                 .sub(tradingFee);
         }
+		if(payload.length==0){
+			address swapImplAddress = IDexRecords(swapsImpl).retrieveDexAddress(1); //default dex address
+			bytes memory dataToSend = abi.encode(sourceToken, destToken);
+			(expectedReturn,) = ISwapsImpl(swapImplAddress).dexAmountOut(
```
Comment (RomanHiden) [inline contracts/swaps/SwapsUser.sol:231]: dexAmountOut can be extracted from if else. see if its possible

File: contracts/swaps/connectors/SwapsImplCurve_ETH.sol [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+contract SwapsImplCurve_ETH is State, ISwapsImpl {
+    using SafeERC20 for IERC20;
+    using Path for bytes;
+    using BytesLib for bytes;
+	address public constant PoolRegistry = 0x63fea6E447F120B8Faf85B53cdaD8348e645D80E; //set to address for monitoring Curve pools
+	bytes4 public constant ExchangeUnderlyingSig = bytes4(keccak256("exchange_underlying(uint256,uint256,uint256,uint256)"));
+	bytes4 public constant ExchangeSig = bytes4(keccak256("exchange(uint256,uint256,uint256,uint256)"));
+	bytes4 public constant GetDySig = bytes4(keccak256("get_dy(uint256,uint256,uint256)"));
+	bytes4 public constant GetDyUnderlyingSig = bytes4(keccak256("get_dy_underlying(uint256,uint256,uint256)"));
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplCurve_ETH.sol:23]: those are constants should be all upper case

File: contracts/mixins/BytesLib.sol [hunk 12]
```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Comment (RomanHiden) [inline contracts/mixins/BytesLib.sol:1]: you could install this lib using `npm install solidity-bytes-utils@0.8.0`

Comment (RomanHiden) [reply contracts/mixins/BytesLib.sol:1]: then add to brownie mapping

File: contracts/mixins/Path.sol [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    function hasMultiplePools(bytes memory path) internal pure returns (bool) {
+        return path.length >= MULTIPLE_POOLS_MIN_LENGTH;
+    }
+
+    /// @notice Returns the number of pools in the path
+    /// @param path The encoded swap path
+    /// @return The number of pools in the path
+    function numPools(bytes memory path) internal pure returns (uint256) {
+        // Ignore the first token address. From then on every fee and token offset indicates a pool.
+        return ((path.length - ADDR_SIZE) / NEXT_OFFSET);
```
Comment (RomanHiden) [inline contracts/mixins/Path.sol:35]: solidity 5.17 doesn't have overflow protection like 8.0 has

**Comment by RomanHiden** [comment]
@Drypto13 can you rebase this pull request to interest-rate code. I guess you will have to resolve some conflicts. interest rates goes out first

File: interfaces/IBZx.sol [hunk 14]
```diff
@@ -406,6 +406,27 @@ interface IBZx {
         uint256 swapAmount, // denominated in collateralToken
         bool returnTokenIsCollateral, // true: withdraws collateralToken, false: withdraws loanToken
         bytes calldata loanDataBytes
+		/* Format:
```
Comment (RomanHiden) [inline interfaces/IBZx.sol:409]: there is a natspec format you can follow to document, [here](https://docs.soliditylang.org/en/latest/natspec-format.html)

**Review by Dpalazue — APPROVED** [review]

File: contracts/farm/FeeExtractAndDistribute_Polygon.sol [hunk 15]
```diff
@@ -3,46 +3,30 @@
  * Licensed under the Apache License, Version 2.0.
  */
 
-pragma solidity 0.6.12;
+pragma solidity 0.8.4;
```
Comment (RomanHiden) [inline contracts/farm/FeeExtractAndDistribute_Polygon.sol:6]: ^0.8.0

File: contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+        internal
+        returns (uint256 sourceTokenAmountUsed, uint256 destTokenAmountReceived)
+    {
+        if (requiredDestTokenAmount != 0) {
+            IUniswapV3SwapRouter.ExactOutputParams[] memory exactParams = abi
+                .decode(payload, (IUniswapV3SwapRouter.ExactOutputParams[]));
+            bytes[] memory encodedTXs = new bytes[](exactParams.length);
+            uint256 totalAmountsOut = 0;
+            uint256 totalAmountsInMax = 0;
+            for (uint256 x = 0; x < exactParams.length; x++) {
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol:155]: rename x to something more understandable

File: contracts/interfaces/ICurvePoolRegistration.sol [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+interface ICurvePoolRegistration {
+    function addPool(
+        address tokenPool,
+        address[] calldata tokensInPool,
+        uint128[] calldata tokenIDs
+    ) external;
+
+    function disablePool(address tokenPool) external;
+
+    function CheckPoolValidity(address pool) external view returns (bool);
```
Comment (RomanHiden) [inline contracts/interfaces/ICurvePoolRegistration.sol:12]: function lowercase

File: contracts/swaps/CurvePoolRegistration/CurvePoolRegistration.sol [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    function addPool(address tokenPool, uint256 PoolT) public {
+        validPool[tokenPool] = true;
+		poolType[tokenPool] = PoolT;
+    }
+
+    function disablePool(address tokenPool) public {
+        validPool[tokenPool] = false;
+    }
+
+    function CheckPoolValidity(address pool) public view returns (bool) {
```
Comment (RomanHiden) [inline contracts/swaps/CurvePoolRegistration/CurvePoolRegistration.sol:18]: lowercase

File: contracts/swaps/SwapsUser.sol [hunk 19]
```diff
@@ -91,7 +95,8 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
         uint256 sourceTokenAmountUsed;
 
         uint256 tradingFee;
-        if (!miscBool) { // bypassFee
+        if (!miscBool) {
```
Comment (RomanHiden) [inline contracts/swaps/SwapsUser.sol:98]: what is miscBool probably need other name for this var

File: contracts/swaps/SwapsUser.sol [hunk 20]
```diff
@@ -124,24 +127,45 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
...
             require(vals[0] <= vals[1], "min greater than max");
         }
-
-        require(loanDataBytes.length == 0, "invalid state");
-        (destTokenAmountReceived, sourceTokenAmountUsed) = _swapsCall_internal(
-            addrs,
-            vals
-        );
-
+        if (loanDataBytes.length == 0) {
```
Comment (RomanHiden) [inline contracts/swaps/SwapsUser.sol:130]: this whole usecase can be optimized
if loanDataBytes.length == 0 && not (flags & DEX_SELECTOR_FLAG != 0) 
normal
else
custom

File: contracts/swaps/SwapsUser.sol [hunk 21]
```diff
@@ -124,24 +127,45 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
...
+                    sourceTokenAmountUsed
+                ) = _swapsCall_internal(addrs, vals, "");
+            }
+        }
         if (vals[2] == 0) {
             // there's no minimum destTokenAmount, but all of vals[0] (minSourceTokenAmount) must be spent, and amount spent can't exceed vals[0]
-            require(sourceTokenAmountUsed == vals[0], "swap too large to fill");
+            require(
+                sourceTokenAmountUsed == vals[0],
+                "swap too large to fill test"
```
Comment (RomanHiden) [inline contracts/swaps/SwapsUser.sol:156]: whats `test`?

File: contracts/swaps/connectors/SwapsImplCurve_ETH.sol [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+            }
+        } else {
+            amountIn = 0;
+        }
+    }
+
+    function dexAmountInFormatted(bytes memory route, uint256 amountOut)
+        public
+        returns (uint256 amountIn, address midToken)
+    {
```
Comment (RomanHiden) [inline contracts/swaps/connectors/SwapsImplCurve_ETH.sol:124]: why not just call return dexAmountIn()

File: contracts/swaps/SwapsUser.sol (hunk 20 above)
Comment (Drypto13) [reply contracts/swaps/SwapsUser.sol:130]: The flags variable is part of loanDataBytes and to get a value for it loanDataBytes has to be decoded which means that if loanDataBytes were to have a length of 0 or not match the formatting expected when being decoded to get flags and bytes[] array of payloads, an error would be thrown as the decode will fail

File: contracts/swaps/SwapsUser.sol (hunk 21 above)
Comment (Drypto13) [reply contracts/swaps/SwapsUser.sol]: Was part of testing dex selector initially, word "test" has been removed.

File: contracts/swaps/SwapsUser.sol (hunk 19 above)
Comment (Drypto13) [reply contracts/swaps/SwapsUser.sol:98]: it is bypassFee I believe. I did not modify the naming of this

File: contracts/interfaces/ICurvePoolRegistration.sol (hunk 17 above)
Comment (Dpalazue) [reply contracts/interfaces/ICurvePoolRegistration.sol]: Thx

**Pull request merged**